Allow routes to be specified as functions

nahiluhmot opened the pull request:
### Anonymous Route Handlers

@hojberg @flahertyb @tlunter @AdamEdgett @barnabyc @ermanc

This pull request allows route handlers to be specified as anonymous functions in the arguments to `Aviator.setRoutes`.
#### Motivation

Aviator's current means of specifying route handlers, via route targets and method names, is useful for grouping the route handling logic of related routes. However, for simple applications it feels a little bit more cumbersome than need be. Consider this application:

``` javascript
var RootTarget = {
  beforeAll: function () {},
  home: function () {},
  about: function () {},
  contact: function () {},
  legal: function () {},
  menu: function () {},
  notFound: function () {}
};

Aviator.setRoutes({
  target: RootTarget,
  '/*': 'beforeAll',
  '/': 'home',
  '/about': 'about',
  '/contact': 'contact',
  '/legal': 'legal',
  '/menu': 'menu',
  notFound: 'notFound'
});

Aviator.dispatch();
```

The above example is not very DRY, since one essentially has to repeat the route name in the arguments to `Aviator.setRoutes` in the handler itself.
#### Description

This PR makes route targets optional by allowing the handlers to be specified as anonymous functions. Using these changes, the above example could be re-written like so:

``` javascript
Aviator.setRoutes({
  '/*': function () {},
  '/': function () {},
  '/about': function () {},
  '/contact': function () {},
  '/legal': function () {},
  '/menu': function () {},
  notFound: function () {}
});

Aviator.dispatch();
```

Although the above example is not modular, one can still break up related routes into separate objects or files. More complex applications may look something like this:

``` javascript
var MarketingRoutes = {
  '/*': function () {},
  '/': function () {},
  notFound: function () {}
};

var StoreRoutes = {
  '/:store': function () {}
};

var BrandRoutes = {
  '/*': function () {},
  '/': function () {},
  '/:brand_id': {
    '/': function () {},
    '/stores': StoreRoutes
  }
  notFound: function () {},
};

var Routes = {
  '/*': function () {},
  '/': function () {},
  notFound: function () {},
  '/marketing': MarketingRoutes,
  '/brands': BrandRoutes,
};

Aviator.setRoutes(Routes);

Aviator.dispatch();
```

Of course, there's no reason that this feature should conflict with the existing Aviator style. Below is an example of an application that uses both the route target/method style and anonymous function style.

``` javascript
var UsersTarget = {
  beforeAll: function () {},
  index: function () {},
  notFound: function () {}
};

var Routes = {
  '/*': function () {},
  '/': function () {},
  notFound: function () {},
  '/users': {
    target: UsersTarget,
    '/*': 'beforeAll',
    '/': 'index',
    notFound: 'notFound',
  }
};

Aviator.setRoutes(Routes);

Aviator.dispatch();
```

There should be good test coverage, but please call out anything I missed.

nahiluhmot committed:
Document the anonymous route handler feature.

The anonymous route handler will allow Aviator consumers to specify
their route handlers as functions. For some applications, this is
preferable to specifying route handlers in a separate object.

nahiluhmot committed:
Add the isFunc predicate function.

nahiluhmot committed:
Add tests for the anonymous route handlers

nahiluhmot committed:
Implement the anonymous actions

nahiluhmot committed:
Regenerate aviator.js

hojberg commented:
I think the current approach of using a RouteTarget is the way we should present as the default option, since it encourages having a separate route file from the implementation (one of the main reasons we built Aviator), but perhaps at least giving the user the choice ok.

I guess the question is how opinionated do we want Aviator to be.

@barnabyc @flahertyb what's your take?

hojberg commented on README.md:
````diff
@@ -129,6 +129,22 @@ Upon hitting `"/marketing/reputation"`,
+```
+
+Upon hitting `"/users/"`, `usersTarget#beforeAll` and the anonymous function
+above will be called in that order.
````
Perhaps add a note that using the targets are encouraged over anonymous functions?

nahiluhmot replied:
Will do.

hojberg commented on aviator.js:
```diff
@@ -840,16 +853,16 @@ Route.prototype = {
         target: target,
         method: routeLevel.notFound
       };
+    } else if (isFunc(routeLevel.notFound)) {
```
move the `else if` below the `}` please

nahiluhmot replied:
My b

hojberg commented on aviator.js:
```diff
@@ -1028,6 +1044,23 @@ Route.prototype = {
+  @return {Object} With a method key that maps to a generated String, and
+  **/
+  anonymousAction: function (func) {
+    var method = 'FUNCTION-' + Object.keys(this.anonymous).length;
```
could we use a timestamp instead? this might collide right?

hojberg replied:
Also do we want to reset `this.anonymous` after all the actions have been run?

nahiluhmot replied:
Wouldn't the timestamp be more likely to collide than this? Being that JS is single-threaded, I don't think this can collide. :+1: on resetting `this.anonymous`.

hojberg replied:
oh i'm sorry i misread this as being based on `func`. Carry on

nahiluhmot replied:
Actually, should we have to reset `this.anonymous`? It should only contain the anonymous matches for one route, which shouldn't be changing.

nahiluhmot replied:
Actually, this could be solved by using a new object for every anonymous function. See [this commit](https://github.com/swipely/aviator/commit/486fdfe3adeedb0ff63ee74307c3b4e8f3874abe).

nahiluhmot commented:
@hojberg Addressed some of your comments in my last two commits, will fix them up if/when we're good to merge.

nahiluhmot committed:
Simplify the anonymous handler implementation.

Instead of having one instance variable called `anonymous` defined on
the `Route`, use a new Object every time.

nahiluhmot commented:
Ping @barnabyc @flahertyb

nahiluhmot commented:
Ping ping @barnabyc @flahertyb

barnabyc commented:
Sorry @nahiluhmot!

barnabyc commented on aviator.js:
```diff
@@ -71,6 +71,16 @@ var isArray = function (o) {
+@private
+**/
+var isFunc = function (val) {
+  return typeof val === 'function';
```
I would append ` || false` so you ensure a bool response.

barnabyc commented:
:+1:  Looks good, but I agree with @hojberg in that the RouteTarget pattern is the preferred, but I can see small projects starting as anonymous and then migrating.

Should also note in the docs that any target will receive an argument of the `request` object.

flahertyb commented:
I totally agree the enforcement of route targets is cumbersome in tiny projects. Dislike complex applications using anonymous functions all over the place, but also dislike unnecessarily opinionated libraries. 

:+1:

nahiluhmot commented:
Sorry about the long silence period. I fixed up the commits and rebased off of master. Good to merge?

hojberg commented:
@nahiluhmot 👍

nahiluhmot commented:
Fast response!
